sw/device/tests/dif/dif_otbn_sanitytest.c

```diff
@@ -0,0 +1,252 @@
+ *
+ * Load the text and data segments into the instruction and data memories,
+ * respectively.
+ *
+ * @param app OTBN application to load
+ */
+static void load_app(const otbn_app_t *app) {
+  CHECK(app->imem_end >= app->imem_start);
+  const size_t imem_size = app->imem_end - app->imem_start;
+
+  CHECK(app->dmem_end >= app->dmem_start);
+  const size_t dmem_size = app->dmem_end - app->dmem_start;
+
+  LOG_INFO(
+      "Loading OTBN instruction memory image stored between address 0x%x and "
+      "%p (%d bytes)",
+      app->imem_start, app->imem_end, imem_size);
+  CHECK(imem_size > 0);
+  CHECK(imem_size % 4 == 0);
+  CHECK(dif_otbn_imem_write(&otbn, 0, (uint32_t *)app->imem_start, imem_size) ==
+            kDifOtbnOk,
+        "Unable to write IMEM application image (.text) to OTBN.");
+
+  if (dmem_size > 0) {
+    LOG_INFO(
+        "Loading OTBN data memory image stored between address %p and %p (%d "
+        "bytes)",
+        app->dmem_start, app->dmem_end, dmem_size);
+    CHECK(dif_otbn_dmem_write(&otbn, 0, (uint32_t *)app->dmem_start,
+                              dmem_size) == kDifOtbnOk,
+          "Unable to write DMEM application image (.data) to OTBN.");
+  } else {
+    LOG_INFO("No OTBN data memory image to load.");
+  }
+}
+
+/**
+ * Call a function on OTBN
+ *
+ * Set the entry point (start address) of OTBN to the desired function, and
+ * starts the OTBN operation.
+ *
+ * @param func the function to be called
+ */
+static void call_function(const otbn_func_t *func) {
+  uint32_t start_address = func->func - func->app->imem_start;
+  LOG_INFO("Calling function at address 0x%x on OTBN.", start_address);
+  CHECK(dif_otbn_start(&otbn, start_address) == kDifOtbnOk);
+}
+
+/**
+ * Busy wait for OTBN to be done with its operation.
+ */
+static void otbn_wait_for_done() {
+  bool busy = true;
+
+  while (busy) {
+    CHECK(dif_otbn_is_busy(&otbn, &busy) == kDifOtbnOk,
+          "Unable to get busy status from OTBN");
+  }
+}
+
+/**
+ * Initialize OTBN's data memory with zeros
+ */
+static void zero_dmem() {
+  const uint32_t zero = 0;
+  int dmem_size_words = dif_otbn_get_dmem_size_bytes(&otbn) / 4;
+  for (int i = 0; i < dmem_size_words; ++i) {
+    dif_otbn_result_t rv =
+        dif_otbn_dmem_write(&otbn, i * 4, &zero, sizeof(zero));
+    CHECK(rv == kDifOtbnOk, "Error zeroing word %d in OTBN DMEM: %d", i, rv);
+  }
+}
+
+/**
+ * Run a 384-bit Barrett Multiplication on OTBN and check its result.
+ *
+ * This test is not aiming to exhaustively test the Barrett multiplication
+ * itself, but test the interaction between device software and OTBN. As such,
+ * only trivial parameters are used.
+ *
+ * The code executed on OTBN can be found in sw/otbn/code-snippets/barrett384.s.
+ * The entry point wrap_barrett384() is called according to the calling
+ * convention described in the OTBN assembly code file.
+ */
+static void test_barrett384() {
+  zero_dmem();
+
+  load_app(&kOtbnAppBarrett384);
+
+  // a, first operand
```

Comment:
These comments say what, not why. Consider removing them.

Reply:
This suggestion seems highly counter-intuitive to me.  

Sometimes (and especially when assembling crypto algo + crypto test code) you want to be explicit and potentially redundant in explaining what you are doing, in case someone comes along later to either fiddle with it, or attempt to leverage it for something else (rightly or wrongly).  To me it seems better to have more clarity than letting someone infer (potentially incorrectly) the wrong intent from the code.

(Unless the concern is about keeping the comment in sync with the code?  That's a bit trickier.)

Reply:
The same variable names are used in the OTBN assembly (`barrett384.s`). For consistency, I'm using the same names here and copied the comments over that describe them. I'm therefore hesitant to rename them, prefix them, or remove the comments.

Reply:
Can you instead provide a comment like "see such and such assembly file for where these variables are used"? I don't think that fact is obvious to a reader.

> Sometimes (and especially when assembling crypto algo + crypto test code) you want to be explicit and potentially redundant in explaining what you are doing

Good comments to not literally repeat what the code below is doing. It's kind of obvious that "a" is the first operand here. A more useful comment might have been "`a` and `b` are the multiplicands for the modular multiplication." Otherwise, I'm left asking "first operand of what?". I don't think that this being security-sensitive cryptographic code affects my observarion.

Reply:
The function refers to the assembly in the comment above it. I don't think this is egregious in this function, given this is test code used in a smoke test. We're allowed to have different standards between test and production code.